### PR TITLE
Check for unsigned_short support for gles2 depth buffer allocation

### LINKED ISSUE #26570: GLES2 error: initialize: Directional shadow framebuffer status invalid on macOS
**Godot version:** 3.1 beta 10

**OS/device including version:** OS X 10.12 AMD Radeon Pro 455

**Issue description:** Hey I noticed when restarting the editor sometimes the old editor doesn't fully close and instead hangs forever. I noticed it more yesterday when I was trying out the GLES2 VS GLES3 settings, but can't reproduce it 100%. Are there logs files somewhere?

**Steps to reproduce:** Hard to reproduce, happens sporadically when doing "Quit to project list" or switching the video driver.

**Minimal reproduction project:** In this case I was testing with Dodge the Creeps.

screen shot 2019-03-04 at 12 56 58 am

### LINKED ISSUE #28894: GLES2 Directional shadow framebuffer status invalid error and shadow bug
**Godot version:**
Version 3.1.1

**OS/device including version:**
Nvidia Geforce 6150SE nforce 430

Windows 7 x64

**Issue description:**
Console Error
image

Shadows ON 
image

Shadow OFF
image

**Steps to reproduce:**
open godot. boom error

for shadow bug. Shadow option is enabled

### PR DESCRIPTION
**Pull request by clayjohn**
Should fix: https://github.com/godotengine/godot/issues/26570 and fix  #28894

Note to reduz: it was as simple as nesting the second check inside the first. The "else" block only gets engaged on mobile when the OES_DEPTH_TEXTURE extension isn't present, so the behaviour can stay the same. On Desktop, we only need to check UNSIGNED_SHORT when UNSIGNED_INT has failed. Then if that fails as well, turn on rgba shadows and set support_depth to false. Internal format is depth16 to match the behaviour on mobile when depth is unsupported.

### CODE REVIEW

File: drivers/gles2/rasterizer_storage_gles2.cpp [hunk 1]
```diff
@@ -5691,21 +5691,49 @@ void RasterizerStorageGLES2::initialize() {
...
+			glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT16, 32, 32, 0, GL_DEPTH_COMPONENT16, GL_UNSIGNED_SHORT, NULL);
+
+			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+			glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+
+			glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, depth, 0);
+
+			GLenum status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
```
Comment (akien-mga) [inline drivers/gles2/rasterizer_storage_gles2.cpp]: Shadows previous local.

**Comment by akien-mga** [comment]
Might also fix quite a few other duplicates of #26570, see referenced issues there.

**Comment by clayjohn** [comment]
Thanks! I took a look through them and the only open one related was https://github.com/godotengine/godot/issues/28894 which should also be closed by this PR.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Zireael07** [comment]
Shouldn't we close #27572, too?

**Comment by endragor** [comment]
The error on OSX was not actually related to UNSIGNED_INT vs UNSIGNED_SHORT. The problem is that GLES2 AMD drivers (any maybe some other) do not support framebuffers without color buffers attached. So, for those drivers `support_depth_texture` should be `true` and for shadow framebuffers a color buffer should be attached.

**Comment by clayjohn** [comment]
@endragor I am aware that ``UNSIGNED_INT`` vs ``UNSIGNED_SHORT`` doesn't matter for OSX. That was the reason for this PR. This PR provides a third branch so that OSX (and others) can fall back on RGBA shadows instead of breaking.

I am concerned about adding a full color buffer to a framebuffer just to access the depth buffer. However, if you think the memory/performance impact would be minor then feel free to make a PR. It would be great to have depth buffer shadows on more devices! :)

**Comment by endragor** [comment]
What I was trying say in my comment is that `support_depth_texture` on those devices shouldn't be set to `false`, while with this change it will be. That flag is used for configuring the main framebuffers later, rather than shadow buffers.

`use_rgba_3d_shadows`, which is the fallback used here, still makes a framebuffer with both depth and color buffers attached, it's just that the depth is drawn into a renderbuffer rather than a texture. So performance-wise I don't think it will be worse if both buffers are textures. It seems we should have 3 options for shadows instead of two:

1. Framebuffer with only depth texture
2. Framebuffer with depth texture and color texture (currently does not exist)
3. Framebuffer with depth renderbuffer and color texture (use_rgba_3d_shadows = true)

**Comment by clayjohn** [comment]
@endragor Ah, I see. Thank you for clarifying. It seems that is an oversight then. 

I don't have an OSX device to test changes on. Do you mind fixing that section up sometime before 3.2? I really would like to have better OSX support for stuff like that. I don't trust myself to get it right just shooting in the dark.

**Comment by akien-mga** [comment]
@endragor @clayjohn If that concern is still valid, it would be good to open a new issue so that we can make sure to fix it for the 3.2 release.

**Pull request merged**